library/lua/gui/widgets.lua

```diff
@@ -1611,6 +1611,15 @@ function List:setChoices(choices, selected)
     end
 
     self:setSelected(selected)
+
+    -- Check if page_top needs to be adjusted
+    if #self.choices - self.page_size < 0 then
+        self.page_top = 1
+    elseif self.selected <= math.floor(self.page_size / 2) then
+        self.page_top = 1
+    elseif self.selected >= #self.choices - math.floor(self.page_size / 2) then
+        self.page_top = #self.choices - self.page_size + 1
+    end
```

Comment:
could you clean this up a little? e.g. #self.choices <= self.page_size is more natural (and I think the <= rather than just < is warranted here)